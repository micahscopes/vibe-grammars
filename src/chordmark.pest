WHITESPACE = _{ " " }

time_signature = { ASCII_NONZERO_DIGIT ~ "/" ~ ASCII_NONZERO_DIGIT }
repeat = { "%" }
duration = { "."* }
chord_for_duration = { (rest | chord) ~ duration }
rest = { "NC" }

end = _{ NEWLINE | EOI }

section_name = { ASCII+ }
section_repeat = { "x" ~ ASCII_DIGIT+ }
key_line = { "key" ~ note ~ end }
time_signature_line = { time_signature ~ end }
section_line = { "#" ~ section_name ~ section_repeat? ~ end}
chord_line = { repeat{0, 2} ~ (time_signature? ~ chord_for_duration+ ~ repeat*)+ ~ end}
lyric_line = { (!NEWLINE ~ ANY)+ ~ end }
song = { (section_line | chord_line | time_signature_line | key_line | lyric_line)+ }