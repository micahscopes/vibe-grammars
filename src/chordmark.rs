use pest_derive::Parser;

#[derive(Parser)]
#[grammar = "elements.pest"]
#[grammar = "chordmark.pest"]
pub struct Parser;