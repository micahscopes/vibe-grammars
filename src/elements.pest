note = { base ~ accidental{, 2} }
base = { "A" | "B" | "C" | "D" | "E" | "F" | "G" }
accidental = { "b" | "#" | "♭" | "♯" | "♮" }


// Chord/scale qualities
long_quality = _{ ^"major" | ^"minor" | ^"diminished" | ^"augmented" | ^"dominant" | ^"altered" }
medium_quality = _{ ^"maj" | ^"min" | ^"dim" | ^"aug" | ^"dom" | ^"alt" }
solfege = { "do" | "re" | "mi" | "fa" | "sol" | "so" | "la" | "ti" }
short_quality = _{ "M" | "m" | "o" | "-" | "+" | "Δ" | "ø" | "°" }
mode = { "lydian" | "mixolydian" | "aeolian" | "locrian" | "ionian" | "dorian" | "phrygian" }

quality = { long_quality | medium_quality | solfege | short_quality }

// Extensions
degree = { "4" | "5" | "6" | "7" | "9" | "10" | "11" | "12" | "13" }
modifier = { "sus" | "add" | "no" | "omit" }
extension = { ((accidental | modifier)? ~ degree) | "alt" }
parenthetical_extension = { "(" ~ extension ~ ")" }

slash_bass = { "/" ~ note }

chord = {
  note ~ quality? ~ (extension | parenthetical_extension)* ~ slash_bass?
}

chords = { chord ~ (" "? ~ chord)* }

key = { note ~ (mode | quality)? }