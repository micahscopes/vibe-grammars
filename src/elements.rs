use pest_derive::Parser;
// #[cfg(debug_assertions)]
// const _GRAMMAR: &'static str = include_str!("chords.pest");

#[derive(Parser)]
#[grammar = "elements.pest"]
pub struct Parser;