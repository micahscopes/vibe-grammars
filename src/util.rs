use pest::iterators::Pairs;
use pest::RuleType;

pub fn to_sexpr<R: RuleType>(pairs: Pairs<R>, indent: usize) -> String {
    let mut output = String::new();
    let indentation = " ".repeat(indent);
    for pair in pairs {
      let rule = pair.as_rule();
      let inner_data = pair.as_str(); // Get the inner data of the pair
      let inner_data = inner_data.replace("\n", &format!("\n{indentation}  "));
      let inner_pairs = pair.into_inner();
      output.push_str(&format!("{indentation}( {rule:?}\n{indentation}  `{inner_data}`\n"));
      output.push_str(&to_sexpr(inner_pairs, indent + 2));
      output.push_str(&format!("{indentation})\n"));
    }
    output
}