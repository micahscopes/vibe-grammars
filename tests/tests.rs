use dir_test::dir_test;
use insta::assert_snapshot;
use pest::{Parser, RuleType};
use vibe_grammars::util::to_sexpr;

fn snapshot_sexpr<P: Parser<R>, R: RuleType>(fixture: dir_test::Fixture<&str>, rule: R) {
    let input = *fixture.content();
    let result = P::parse(rule, input).unwrap();
    let tree = to_sexpr(result, 0);
    let stem = std::path::Path::new(fixture.path()).file_stem().unwrap().to_str().unwrap();
    assert_snapshot!(stem, tree);
}

use vibe_grammars::chordmark::{Parser as ChordMarkParser, Rule as ChordMarkRule};
use vibe_grammars::elements::{Parser as ElementsParser, Rule as ElementsRule};
#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/tests/elements",
    glob: "*.txt",
)]
fn test_elements(fixture: dir_test::Fixture<&str>) {
  snapshot_sexpr::<ElementsParser, ElementsRule>(fixture, ElementsRule::chords);
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/tests/chordmark",
    glob: "*.txt",
)]
fn test_chordmark_song(fixture: dir_test::Fixture<&str>) {
  snapshot_sexpr::<ChordMarkParser, ChordMarkRule>(fixture, ChordMarkRule::song);
}