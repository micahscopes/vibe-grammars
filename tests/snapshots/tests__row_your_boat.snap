---
source: tests/tests.rs
assertion_line: 11
expression: tree
---
( song
  `G.    Em.   G..
  Row, row, row your boat
  G.      Em.       D..
  Gently down the stream
  G.        D.        G.        Em.
  Merrily, merrily, merrily, merrily
  D7..            G..
  Life is but a dream`
  ( chord_line
    `G.    Em.   G..
    `
    ( chord_for_duration
      `G.`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `Em.`
      ( chord
        `Em`
        ( note
          `E`
          ( base
            `E`
          )
        )
        ( quality
          `m`
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `G..`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `..`
      )
    )
  )
  ( lyric_line
    `Row, row, row your boat
    `
  )
  ( chord_line
    `G.      Em.       D..
    `
    ( chord_for_duration
      `G.`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `Em.`
      ( chord
        `Em`
        ( note
          `E`
          ( base
            `E`
          )
        )
        ( quality
          `m`
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `D..`
      ( chord
        `D`
        ( note
          `D`
          ( base
            `D`
          )
        )
      )
      ( duration
        `..`
      )
    )
  )
  ( lyric_line
    `Gently down the stream
    `
  )
  ( chord_line
    `G.        D.        G.        Em.
    `
    ( chord_for_duration
      `G.`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `D.`
      ( chord
        `D`
        ( note
          `D`
          ( base
            `D`
          )
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `G.`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `.`
      )
    )
    ( chord_for_duration
      `Em.`
      ( chord
        `Em`
        ( note
          `E`
          ( base
            `E`
          )
        )
        ( quality
          `m`
        )
      )
      ( duration
        `.`
      )
    )
  )
  ( lyric_line
    `Merrily, merrily, merrily, merrily
    `
  )
  ( chord_line
    `D7..            G..
    `
    ( chord_for_duration
      `D7..`
      ( chord
        `D7`
        ( note
          `D`
          ( base
            `D`
          )
        )
        ( extension
          `7`
          ( degree
            `7`
          )
        )
      )
      ( duration
        `..`
      )
    )
    ( chord_for_duration
      `G..`
      ( chord
        `G`
        ( note
          `G`
          ( base
            `G`
          )
        )
      )
      ( duration
        `..`
      )
    )
  )
  ( lyric_line
    `Life is but a dream`
    ( EOI
      ``
    )
  )
)
